### ISSUE TITLE
Can the dependency of FFTW be replaced by AbstractFFTs?

### ISSUE THREAD

**Issue by heliosdrm** [bionanoimaging/FourierTools.jl#50]
FFT.jl is a particular implementation of AbstractFFTs.jl, with a few exensions documented in https://juliamath.github.io/FFTW.jl/stable/fft/
But unless I have missed something, FourierTools does not use any of those functions.

So perhaps the dependency from FFTW might be dropped in favor of AbstractFFTs, and let users choose their preferred implementation (FFTW if they want, but also maybe [FastTransforms](https://github.com/JuliaApproximation/FastTransforms.jl), or [RustFFT](https://github.com/Taaitaaiger/RustFFT.jl), for instance).

This might be arguably breaking, since users should install any of those implementations explicitly in order to use this package ~(in the fashion of FileIO)~, although nothing of their code would change. The advantages are that FourierTools would be lighter and more flexible.

**Comment by roflmaostc** [comment]
Good remark.
What happens to all internal functions of FourierTools which call `fft`?

So to effectively use FourierTools, user always need to load a FFT backend. Right?

**Comment by RainerHeintzmann** [comment]
Maybe one can ship it with "batteries", i.e. FFTW included but it then uses the FourierToolsCore package where it exports also almost all it's functions?
This was we can maintain compatibility.

**Comment by roflmaostc** [comment]
Yeah we could but it would be some extra work.

So not sure if it's worth it.


Are the other mentioned packages in any way competitive with FFTW.jl already? So is it really of practical need?

**Comment by heliosdrm** [comment]
FFTW.jl has the drawback of relying on a third party library with a GPL license, and in some specific situations might be a barrier to depend on it. FastTransforms is pure Julia, and RustFFT depends on a library licensed under Apache and MIT. RustFFT.jl also claims to be faster than FFTW.jl (see its [README](https://github.com/Taaitaaiger/RustFFT.jl)), although it is limited to one-dimensional variables.

**Comment by heliosdrm** [comment]
> So to effectively use FourierTools, user always need to load a FFT backend. Right?

Yes. e.g. `using FourierTools` (or `using FourierToolsCore` as per @RainerHeintzmann's suggestion) would not work alone. It should be `using FourierToolsCore, FFTW` or similar.

**Comment by roflmaostc** [comment]
Yeah, or we would do two packages
FourierToolsCore
And
FourierTools


The latter would pull in FFTW.jl as default.

**Comment by heliosdrm** [comment]
> FastTransforms is pure Julia

Sorry, this statement was wrong: FastTransforms is a wrapper to a C library (MIT licensed).

**Comment by RainerHeintzmann** [comment]
Are you sure that FFTW.jl is not free to use in Julia?
At least the package is MIT licenced. 
https://github.com/JuliaMath/FFTW.jl/blob/master/LICENSE
That is not GPL.

**Comment by roflmaostc** [comment]
Tdlr; The Julia code of FFTW.jl is MIT but not underlying FFTW code. 

See https://github.com/JuliaMath/FFTW.jl?tab=readme-ov-file#license

**Comment by heliosdrm** [comment]
I was going to prepare a PR to test how this proposed change would look like, and have realized that it is not totally straightforward, because there are a few functions (`plan_conv` plus others in `czt.jl`) that do depend on some constants defined in FFTW.

**Comment by RainerHeintzmann** [comment]
... and possibly fftshift?  Or is that part of the general framework?

**Comment by roflmaostc** [comment]
This is covered by AbstractFFTs.

**Comment by RainerHeintzmann** [comment]
@heiosdrm Any progress on this? Maybe we could somehow use Julias Extension mechanism to define which FFT is used?

**Comment by heliosdrm** [comment]
No, sorry, I stopped my attempt when I realized that:

>  there are a few functions (`plan_conv` plus others in `czt.jl`) that do depend on some constants defined in FFTW.

Making changes to solve that kind of conflicts was beyond the scope of my initial intention.
